modules/caddytls/ondemand.go

```diff
@@ -0,0 +1,192 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caddytls
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/caddyserver/certmagic"
+	"go.uber.org/zap"
+
+	"github.com/caddyserver/caddy/v2"
+)
+
+func init() {
+	caddy.RegisterModule(PermissionByHTTP{})
+}
+
+// OnDemandConfig configures on-demand TLS, for obtaining
+// needed certificates at handshake-time. Because this
+// feature can easily be abused, you should use this to
+// establish rate limits and/or an internal endpoint that
+// Caddy can "ask" if it should be allowed to manage
+// certificates for a given hostname.
+type OnDemandConfig struct {
+	// DEPRECATED. WILL BE REMOVED SOON. Use 'permission' instead.
+	Ask string `json:"ask,omitempty"`
+
+	// REQUIRED. A module that will determine whether a
+	// certificate is allowed to be loaded from storage
+	// or obtained from an issuer on demand.
+	PermissionRaw json.RawMessage `json:"permission,omitempty" caddy:"namespace=tls.permission inline_key=module"`
+	permission    OnDemandPermission
+
+	// DEPRECATED. An optional rate limit to throttle
+	// the checking of storage and the issuance of
+	// certificates from handshakes if not already in
+	// storage. WILL BE REMOVED IN A FUTURE RELEASE.
+	RateLimit *RateLimit `json:"rate_limit,omitempty"`
+}
+
+// DEPRECATED. WILL LIKELY BE REMOVED SOON.
+// Instead of using this rate limiter, use a proper tool such as a
+// level 3 or 4 firewall and/or a permission module to apply rate limits.
+type RateLimit struct {
+	// A duration value. Storage may be checked and a certificate may be
+	// obtained 'burst' times during this interval.
+	Interval caddy.Duration `json:"interval,omitempty"`
+
+	// How many times during an interval storage can be checked or a
+	// certificate can be obtained.
+	Burst int `json:"burst,omitempty"`
+}
+
+// OnDemandPermission is a type that can give permission for
+// whether a certificate should be allowed to be obtained or
+// loaded from storage on-demand.
+// EXPERIMENTAL: This API is experimental and subject to change.
+type OnDemandPermission interface {
+	// CertificateAllowed returns nil if a certificate for the given
+	// name is allowed to be either obtained from an issuer or loaded
+	// from storage on-demand.
+	//
+	// The context passed in has the associated *tls.ClientHelloInfo
+	// value available at the certmagic.ClientHelloInfoCtxKey key.
+	//
+	// In the worst case, this function may be called as frequently
+	// as every TLS handshake, so it should return as quick as possible
+	// to reduce latency. In the normal case, this function is only
+	// called when a certificate is needed that is not already loaded
+	// into memory ready to serve.
+	CertificateAllowed(ctx context.Context, name string) error
```

Comment:
What do you think about adding the issuer key that is requesting the cert here?  For some use cases, I would want to allow or reject the issuance based on who is asking.

Reply:
I don't think an _issuer_ has been determined yet. Multiple issuers can be configured (for example, for redundancy), and if one fails it'll try the next, etc. So this function is called before trying any issuers, and is called only once.

But when you say:

> I would want to allow or reject the issuance based on who is asking.

To clarify, Issuers are entities that issue certificates, not entities that ask for a certificate. The server is asking (or, depending on how you look at it, the client is asking for a cert). Can you elaborate what your needs are?

Reply:
There's no such thing, the issuer is not involved at this point. This is before we start issuance, it's for us to know "should we do issuance?" and then if we get a yes it'll choose an issuer.

Reply:
I recall the request flow now, so what you are saying makes sense.

Basically I wanted the ask endpoint to, in effect, be per-issuer.  I know you can filter what issuer is appropriate by using the `subjects` param, but the list of what's appropriate is external to config.  It can be placed behind the `OnDemandPermission` but that's global for all issuers, as noted.

What's in this PR works for my exact use case as long as I have one issuer -- which is the case.  So probably not that big of a deal.

Reply:
Yeah, that starts rolling into "dynamically load the TLS config" territory. Which is probably possible, but neither efficient nor simple. Maybe let's revisit that if needed in a separate issue...